### PR TITLE
doc: reference: terminology: refine API attributes

### PR DESCRIPTION
**Pull request by pabigot**
The original description for isr-ok allowed the function to return an
error if called from an interrupt context, rather than doing its job.
This led to complex descriptions like "isr-ok but not sleep" to
indicate that the function could be called from thread or interrupt
context without visible behavior changes, including returning an
error based on calling context.

Rename the isr-ok (pre-kernel-ok) attribute to isr-safe
(pre-kernel-safe) to specify that calls from non-thread context will
not cause undefined behavior (i.e. are safe).  Redefine isr-ok to
indicate that the call does what it's supposed to do regardless of
interrupt or thread context.

This also makes more clear that isr-ok applies to functions regardless
of whether they can cause an invoking thread to context switch.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

### CODE REVIEW

**Comment by hakonfam** [comment]
I would rename "isr-ok" to something like "isr-functional/correct".
To me the "isr-ok" could provide a better description of its meaning in its name.
IMO there is little difference between "ok" and "safe", and "isr-ok" really means that it is fully functional when called from this context, not only "ok". 

I'm not an english language expert, so I might be wrong on this.

Since none of these terms are in use as of yet (as for as I know of) I guess we can still change them to make them more expressive.

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by pabigot** [comment]
I don't think `isr-functional` is nearly as clear as `isr-ok`, so I'm proposing another path:

I originally wanted a way to make clear that invoking a function from interrupt context would not cause the system to crash, but might do so by returning an error.  Now I need a way to make clear that invoking a function produces the same result whether in interrupt or thread context; returning an error is *not* an acceptable way to avoid crashes.

I propose to reconcile these by specifying **foo-ok** to mean what people would normally expect, and delegate documentation of how it's accomplished to the individual API.  Thus some functions will document that they'll return `-EWOULDBLOCK` when invoked from interrupt (or pre-kernel), while others will do what they're supposed to do, but nothing that's **isr-ok** is allowed to crash the system when invoked from interrupt context.

DNM until "enough" time has passed to allow people to examine both options (currently in two commits) to see if the original approach has any merit.  If it doesn't, they'll be squashed.

**Review by carlescufi — APPROVED** [review]
LGTM

**Comment by carlescufi** [comment]
@pabigot please squash

File: doc/reference/terminology.rst [hunk 1]
```diff
@@ -122,32 +123,33 @@ pre-kernel contexts only when the parameter selects the no-wait path.
 isr-ok
 ======
 
-The isr-ok attribute is used on a function to indicate that it can be
-called from interrupt context.  If necessary the function will use
-:cpp:func:`k_is_in_isr` to detect its calling context and force an
-execution path that will not cause the invoking thread to sleep.
+The isr-ok attribute is used on a function to indicate that it works as
```
Comment (nashif) [inline doc/reference/terminology.rst:126]: I think this is an awkward wording that ends up not adding much value. "indicate that it works as documented....". Why do we need this if we have to rely on the documentation, IMO the original text was more to the point: functions with this attribute can be called from interrupt context, this was straight and to the point.

**Review by nashif — CHANGES_REQUESTED** [review]

File: doc/reference/terminology.rst (hunk 1 above)
Comment (pabigot) [reply doc/reference/terminology.rst]: ok

**Review by nashif — APPROVED** [review]

**Pull request merged**